orderer/common/filerepo/filerepo.go

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright IBM Corp. All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package filerepo
+
+import (
+	"bufio"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/hyperledger/fabric/common/configtx"
+)
+
+type Repo struct {
+	Options Options
+}
+
+type Options struct {
+	FileSuffix          string
+	TransientFilePrefix string
+	FileRepoDir         string
+}
+
+// Save atomically persists the content to the channelName+suffix file by first writing it
+// to a tmp file prefixed by the TransientFilePrefix and then moves the file to the final
+// destination indicated by the FileSuffix.
+func (r *Repo) Save(channelName string, content []byte) error {
+	err := configtx.ValidateChannelID(channelName)
+	if err != nil {
+		return err
+	}
+
+	fileRepoDir := r.Options.FileRepoDir
+	dest := filepath.Join(fileRepoDir, channelName+r.Options.FileSuffix)
+
+	tmpFile, err := ioutil.TempFile(fileRepoDir, r.Options.TransientFilePrefix+channelName)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if err != nil {
+			tmpFile.Close()
+			os.Remove(tmpFile.Name())
+		}
+	}()
+
+	w := bufio.NewWriter(tmpFile)
+
+	if _, err := w.Write(content); err != nil {
+		return err
+	}
+
+	if err := w.Flush(); err != nil {
+		return err
+	}
+
+	if err := tmpFile.Close(); err != nil {
+		return err
+	}
+
+	return os.Rename(tmpFile.Name(), dest)
+}
+
+// Scan parses the directory and produce a list of file names, filtered by suffix.
+func (r *Repo) Scan() ([]string, error) {
```

Comment:
Nit. `List()` may be more expressive.

Reply:
I don't think these functions would ever be called concurrently as saving occurs at the rest API level on receiving a join channel request; scan and load are called in order on orderer startup/registrar initialization to trigger onboarding for each join block.

Reply:
I see. I thought of this package as a generic file repo that has no assumptions about how the caller would use it. With your explanation, my assumption no longer holds and I don't know whether non-generic pkgs can reside in the common.

Reply:
If the API doesn't state it is thread safe the assumption of the user should be that it's not

Reply:
Hmm good point. I think we may need to take a step back and evaluate whether this should be designed around general purpose usage or specifically for channel participation right now as a lot of the discussions below seem to hinge on this nuance. There's always the option of keeping it as specific to channel participation for now and in the future if we ever need to make it more general purpose, some small changes here and there could be included at that time. It's easy to say this could be used for other purposes, but right now I'm blanking on thinking of exactly what else we could use this for.

Reply:
Adding a lock and making it thread-safe is very easy to do it for this pkg. Correct me If I am missing any challenges. That's why I suggested and I believe that making these kinds of packages generic would help in a bug-free code (in the future). Just take it as my philosophical view and I would better leave it to the orderer squad to choose what is best. At least in the ledger, it helps us and we don't need to check everytime whether the call we make is thread safe or not :-)